lib/puppet/type/zfs_share.rb

```diff
@@ -0,0 +1,141 @@
+Puppet::Type.newtype(:zfs_share) do
+  @doc = "Manage zfs shares"
+
+  require 'puppet/parameter/boolean'
+
+  ensurable
+
+  newparam(:zfs_name) do
+    desc "Name of zfs file system"
+    isnamevar
+
+    validate do |value|
+      unless value =~ /(?!\/|\d)^\w.*/
+        raise ArgumentError , "%s is not a valid zfs path."
+      end
+    end
+  end
+
+  newparam(:zpool) do
+    desc "Name of the zfs parent"
+
+    validate do |value|
+      unless value =~ /(?!\/|\d)^\w.*/
+        raise ArgumentError , "%s is not a valid zpool."
+      end
+    end
+  end
+
+  newparam(:share_title) do
+    desc "Override the title of the share"
+
+    validate do |value|
+      if value =~ /\s/
+        raise ArgumentError , "#{value} is not a valid share title"
+      end
+    end
+  end
+
+  newparam(:allow_ip_read, :array_matching => :all) do
```

Comment:
How do you mean? Read and write IPs are both optional. I don't want the user to have to specify the entire read/write string. It seems more puppet like to accept an array of IP addresses and map them, though I could also accept a hash of read and write addresses within an allow_ip parameter. A user can set * to allow all addresses if desired. My goal here is to make sure that action is explicit and doesn't happy unintentionally.